lazyClient.go

```diff
@@ -147,6 +147,22 @@ func (l *lazyClientConn[T]) Close() error {
 	// closed the stream for reading. I mean, we're the initiator so that's
 	// strange... but it's still allowed
 	_ = l.Flush()
+
+	// Finish reading the handshake before we close the connection/stream. This
+	// is necessary so that the other side can finish sending its response to our
+	// multistream header before we tell it we are done reading.
+	//
+	// Example:
+	// We open a QUIC stream, write the protocol `/a`, send 1 byte of application
+	// data, and immediately close.
+	//
+	// This can result in a single packet that contains the stream data along
+	// with a STOP_SENDING frame. The other side may be unable to negotiate
+	// multistream select since it can't write to the stream anymore and may
+	// drop the stream.
+	//
+	// Note: We currently handle this case in Go, but rust-libp2p does not.
+	l.rhandshakeOnce.Do(l.doReadHandshake)
```

Comment:
So the read handshake happened in both `Flush` and `Write`, but it was done asynchronously(loc 172 and 128), and the connection could be closed before readHandshake finishes.... Makes sense

Reply:
Btw, the comment on `Close` states no flushing is happening, but it actually does above.

Reply:
Also, as lazy wrapper always Flushes on close, do we actually need another wrapper in basichost https://github.com/libp2p/go-libp2p/blob/7268c98442c34d26a5d87cd72e37c48e1ffe2e6c/p2p/host/basic/basic_host.go#L1151-L1161?

Reply:
Thanks for pointing that out. I've been meaning to remove it. I'm not sure why we need to wrap the whole thing at all.

Reply:
I guess it did at some point. Go clients handle this with: #87

Reply:
> Also, as lazy wrapper always Flushes on close, do we actually need another wrapper in basichost

I think yes because we want to flush on CloseWrite as well.